frontend/src/pages/new_question/utils/queryBuilderUtils.js

```diff
@@ -0,0 +1,199 @@
+    // loop over all connected nodes to find edges that are connected to them
+    [...connectedNodes].forEach((nodeId) => {
+      [...disconnectedEdges].forEach((e) => {
+        const edge = q_graph.edges[e];
+        // if edge is connected to node
+        if (edge.subject === nodeId || edge.object === nodeId) {
+          connectedNodes.add(edge.subject);
+          connectedNodes.add(edge.object);
+          disconnectedEdges.delete(e);
+          foundConnection = true;
+        }
+      });
+    });
+    // no more connected edges, break out
+    if (!foundConnection) {
+      foundConnections = false;
+    }
+  }
+  q_graph.edges = _.omitBy(q_graph.edges, (e, id) => disconnectedEdges.has(id));
+  // delete all floating nodes
+  const notFloatingNodeIDs = new Set();
+  Object.values(q_graph.edges).forEach((e) => {
+    notFloatingNodeIDs.add(e.subject);
+    notFloatingNodeIDs.add(e.object);
+  });
+
+  // Trim a node if it is floating and marked for deletion
+  const nodesToTrim = Object.keys(q_graph.nodes).filter((id) => (!notFloatingNodeIDs.has(id) && id !== rootNode));
+
+  q_graph.nodes = _.omitBy(q_graph.nodes, (n, id) => nodesToTrim.includes(id));
+  // q_graph.nodes = _.pick(q_graph.nodes, notFloatingNodeIDs);
+  return q_graph;
+}
+
+/**
+ * After deleting a node, trim any edges connected to it
+ * @param {object} q_graph query graph
+ * @param {string} deletedNode deleted node id
+ * @param {string} rootNode root node id
+ * @returns trimDetached query graph
+ */
+function trimDetachedEdges(q_graph, deletedNode, rootNode) {
+  const edgeIds = Object.keys(q_graph.edges).map((id) => id);
+  edgeIds.forEach((eId) => {
+    const currentEdge = q_graph.edges[eId];
+    if (currentEdge.subject === deletedNode || currentEdge.object === deletedNode) {
+      delete q_graph.edges[eId];
+    }
+  });
+  let root = rootNode;
+  if (root === deletedNode) {
+    root = findRootNode(q_graph);
+  }
+  return trimDetached(q_graph, root);
+}
+
+/**
+ * Get a unique list of all edges connected to specified node
+ * @param {object} edges query graph edges object
+ * @param {string} nodeId node id
+ * @returns Set of edge ids
+ */
+function getConnectedEdges(edges, nodeId) {
+  return new Set(Object.keys(edges).filter((eId) => (
+    edges[eId].subject === nodeId || edges[eId].object === nodeId
+  )));
+}
+
+function computeRootNode(q_graph, rootNode) {
+  if (getConnectedEdges(q_graph.edges, rootNode).size > 0) {
+    return rootNode;
+  }
+  return findRootNode(q_graph);
+}
+
+/**
+ * Starting at root node, is this query graph valid? Is there at least one hop in the graph?
+ * @param {object} q_graph query graph object
+ * @param {string|undefined} rootNode node id
+ * @returns boolean
+ */
+function isValidGraph(q_graph, rootNode) {
+  const clonedQueryGraph = _.cloneDeep(q_graph);
+  let isValid = false;
+  let root = rootNode;
+  while (!isValid) {
+    if (root === undefined) {
+      root = findRootNode(clonedQueryGraph);
```

Comment:
Again, I don't think we should be changing the root node here, unless we re-conceptualize the role of this function.

Reply:
Refactored in 006493f40e1cf2d052b713a5c5ae8d108a18540d